src/postgres.rs

```diff
@@ -0,0 +1,275 @@
+use crate::db::{Db, DbConnection};
+use actyxos_sdk::{
+    event::{Event, SourceId, StreamInfo},
+    LamportTimestamp, Offset, OffsetMap, Payload,
+};
+use anyhow::Result;
+use async_trait::async_trait;
+use serde_json::Value;
+use std::{
+    collections::BTreeMap,
+    convert::{TryFrom, TryInto},
+    time::Instant,
+};
+use tokio_postgres::{types::Type, NoTls};
+use tokio_postgres::{Client, Statement};
+use tracing::{debug, info};
+
+#[derive(Debug, PartialEq)]
+pub struct DbEventVec<'a> {
+    pub sources: Vec<&'a str>,
+    pub semantics: Vec<&'a str>,
+    pub names: Vec<&'a str>,
+    pub lamports: Vec<i64>,
+    pub offsets: Vec<i64>,
+    pub timestamps: Vec<i64>,
+    pub payloads: Vec<Value>,
+}
+
+impl<'a> DbEventVec<'a> {
+    pub fn empty(capacity: usize) -> DbEventVec<'a> {
+        DbEventVec {
+            sources: Vec::with_capacity(capacity),
+            semantics: Vec::with_capacity(capacity),
+            names: Vec::with_capacity(capacity),
+            lamports: Vec::with_capacity(capacity),
+            offsets: Vec::with_capacity(capacity),
+            timestamps: Vec::with_capacity(capacity),
+            payloads: Vec::with_capacity(capacity),
+        }
+    }
+}
+
+impl<'a> From<&'a Vec<Event<Payload>>> for DbEventVec<'a> {
+    fn from(events: &'a Vec<Event<Payload>>) -> DbEventVec<'a> {
```

Comment:
might be just the right thing here — in more general cases I’d take a `&'a[Event<Payload>]`, which can easily be obtained from a `v: Vec` using `&*v`.

Reply:
cool... fixed